ateam_ai/src/plays/defense.hpp

```diff
@@ -0,0 +1,125 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#ifndef PLAYS__DEFENSE_HPP_
+#define PLAYS__DEFENSE_HPP_
+
+#include <Eigen/Dense>
+#include <array>
+#include <vector>
+#include <optional>
+
+#include "types/world.hpp"
+#include "types/behavior_goal.hpp"
+#include "ateam_geometry/utilities.hpp"
+#include "ateam_geometry/ateam_geometry.hpp"
+
+const FieldSidedInfo & our_side_info
+DirectedGraph<BehaviorGoal> generate_basic_defense(
+  const World & world,
+  const Feild & field,
+  const FieldSidedInfo & our_side_info)
+{
+  DirectedGraph<BehaviorGoal> defense_graph;
+  // Go to the middle of the goalie area
+  BehaviorGoal goalie = get_goalie_behavior_goal(our_side_info);
+  defense_graph.add_node(goalie);
+
+  std::vector<BehaviorGoal> defenders = get_defense_behavior_goals(world, field, 2);
+  for (BehaviorGoal defender : defenders) {
+    defense_graph.add_node(defender);
+  }
+  return defense_graph;
+}
+
+BehaviorGoal get_goalie_behavior_goal(const FieldSidedInfo & our_side_info)
+{
+  // Line that is 0.5 m from the defense area in all directions
+  ateam_geometry::Segment goalie_line = ateam_geometry::Segment(
+    ateam_geometry::Point(
+      -4,
+      0.5), ateam_geometry::Point(
+      -4, -0.5));
+
+  // Get the ball location
+  std::optional<Eigen::Vector2d> ball_location = world.get_unique_ball();
+  while (!ball_location.has_value()) {
+    ball_location = world.get_unique_ball();
+  }
+
+  // Get the point on the goalie line that is closest to the ball
+  ateam_geometry::Point _goalie_point = goalie_line.projection(ball_location);
+
+  // Have the goalie defend the goal by going to that point
+  return BehaviorGoal goalie {
+    BehaviorGoal::Type::MoveToPoint,
+    BehaviorGoal::Priority::Required,
+    MoveParam(_goalie_point)
+  };
+}
+
+std::vector<BehaviorGoal> get_defense_behavior_goals(
+  const World & world, const Field & field,
+  const int & num_defenders)
+{
+  std::vector<BehaviorGoal> defenders;
+  std::optional<Eigen::Vector2d> ball_location = world.get_unique_ball();
+  // Get line between the ball and the goal
+  while (!ball_location.has_value()) {
+    ball_location = world.get_unique_ball();
+  }
+  ateam_geometry::Point ball = ateam_geometry::EigenToPoint(ball_location);
```

Comment:
Same with the whole loop here

Reply:
Addressed this in the same way as the other one